languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java

```diff
@@ -2339,6 +2354,231 @@ private boolean isOnlyPluralNoun(String word) throws IOException {
+
+    // For some part1-and-part2 combinations an infix s is correct or incorrect
+    if (needsSometimesInfixS(part1, part2)) {
+      return false;
+    }
+
+    // TODO distinguish more cases with hyphens
+    if (part2ucIsNoun && !part2ucIsMisspelled &&
+      // 's' is the last character in *part1* and is probably not an infix
+      part1WithoutHyphen.endsWith("s") && (isNounNom(part1uc) || isVerbStem(part1)) &&
+      // check if infix 's' is required or not allowed
+      (!hasNoInfixS(removeInfixS(part1uc)) || needsInfixS(removeInfixS(part1uc)))) {
+      return true;
+    }
+    if (part2ucIsNoun && !part2ucIsMisspelled &&
+      // 's' is the last character in *part1* and is probably an infix
+      part1WithoutHyphen.endsWith("s") && isNounNom(removeInfixSAndHyphen(part1uc)) &&
+      // check if infix 's' is required or not allowed
+      (!hasNoInfixS(removeInfixSAndHyphen(part1uc)) || needsInfixS(removeInfixSAndHyphen(part1uc)))) {
+      return true;
+    }
+    if (part2ucIsNoun && !part2ucIsMisspelled &&
+      // *part1* does not end with 's' and is noun or verb stem
+      (!part1WithoutHyphen.endsWith("s")) && (isNounNom(part1uc) || isVerbStem(part1)) &&
+      // check if infix 's' is required or not allowed
+      (hasNoInfixS(part1uc) || !needsInfixS(part1uc))) {
+      return true;
+    }
+    if (part2ucIsNoun && !part2ucIsMisspelled &&
+      // *part1* is acronym, e. g. "SEO-Expertinnen"
+      isAllUppercase(removeInfixSAndHyphen(part1)) && !isMisspelled(removeInfixSAndHyphen(part1))) {
+      return true;
+    }
+    return false;
+  }
+
+  protected boolean processThreePartCompound(List<String> parts) throws IOException {
+    String part1 = parts.get(0);
+    String part2 = parts.get(1);
+    String part3 = parts.get(2);
+    String compound1 = part1 + part2;
+    String compound2 = uppercaseFirstChar(part2) + parts.get(2);
+
+    if (isNoun(compound1) && isNoun(compound2)) {
+      // If part1part2 and part2part3 are compounds, then so is part1part2part3
+      return (processTwoPartCompounds(part1, removeHyphen(part2)) && processTwoPartCompounds(part2, part3));
+    }
+    if (compound1.endsWith("s") || compound1.endsWith("s-")) {
+      String part2NoInfixSNoHyphen = removeInfixSAndHyphen(part2);
+      // If part1part2NoInfixSNoHyphen and part2part3 are compounds, then so is part1part2part3
+      return (processTwoPartCompounds(part1, part2NoInfixSNoHyphen) && processTwoPartCompounds(part2, part3));
+    }
+    if (isVerbPrefix(part1) && isVerbStem(part2) && isNoun(compound2)) {
+      // e.g. "Aus" + "leih" + "stelle"
+      return true;
+    }
+    if (isNounNom(part1) && isVerbStem(part2) && isNoun(compound2)) {
+      // e.g. "Wein" + "kühl" + "schrank"
+      return true;
+    }
+    return false;
+  }
+
+  private boolean needsSometimesInfixS (String part1, String part2) throws IOException {
+    // Check if *part1* requires or prohibits infix s depending on *part2*
```

Comment:
❓ The name of the method and the descriptive comment are a little confusing to me.

If you pass `Arbeit` and `platz`, how does it **sometimes** need a Fugen-S?

I understand what the method **does**, but I think the name and the comment (as well as the comment when it's actually called) don't really make it apparent to someone reading the code immediately.